paddle/fluid/framework/ir/conv_elementwise_add_mkldnn_fuse_pass.cc

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/framework/ir/conv_elementwise_add_mkldnn_fuse_pass.h"
+#include <functional>
+
+#include "paddle/fluid/framework/ir/graph_traits.h"
+
+namespace paddle {
+namespace framework {
+namespace ir {
+namespace patterns {
+
+template <typename IT, typename FindFunc, typename ReplaceFunc>
+static void ReplaceAllOccurances(IT s, IT e, FindFunc f, ReplaceFunc r) {
+  if (s == e) return;
+
+  auto it = std::find_if(s, e, f);
+
+  if (it != e) {
+    r(*it);
+  }
+
+  it++;
+  ReplaceAllOccurances(it, e, f, r);
+}
+
+static void CorrectGraphEdges(Graph* graph, Node* from, Node* to) {
+  for (auto& node : GraphTraits::DFS(*graph)) {
+    auto same = std::find_if(std::begin(node.inputs), std::end(node.inputs),
+                             [from](Node* n) { return n == from; });
+
+    if (same != std::end(node.inputs)) {
+      IR_NODE_LINK_TO(to, (&node));
+
+      auto inputs = node.Op()->Inputs();
+
+      using input_type = VariableNameMap::value_type;
+
+      ReplaceAllOccurances(
+          std::begin(inputs), std::end(inputs),
+          [from](const input_type& i) -> bool {
+            auto params = i.second;
+            auto pi =
+                std::find_if(std::begin(params), std::end(params),
+                             std::bind(std::equal_to<std::string>(),
+                                       from->Name(), std::placeholders::_1));
+            return pi != std::end(params);
+          },
+          [to, &node](const input_type& i) {
+            node.Op()->SetInput(i.first, {to->Name()});
+          });
+    }
+  }
+}
+}  // namespace patterns
+using graph_ptr = std::unique_ptr<ir::Graph>;
+
+graph_ptr ConvElementwiseAddMKLDNNFusePass::ApplyImpl(graph_ptr graph) const {
+  FusePassBase::Init("conv_elementwise_add_mkldnn_fuse_pass", graph.get());
+
+  GraphPatternDetector gpd;
+  auto pattern = gpd.mutable_pattern();
+
+  patterns::Conv conv_pattern{pattern, "skip_connections_fusion"};
+  auto conv_output = conv_pattern();
+
+  patterns::ElementwiseAdd elementwise_add_pattern{pattern,
+                                                   "skip_connections_fusion"};
+  elementwise_add_pattern(conv_output);
```

Comment:
> How can the case when convolution output could be used as either `X` or `Y` argument be implemented?

I think you can pass additional argument to this function function, a bool for example, which determines whether it should be X or Y. Then add an if() inside this function.

Reply:
Adding a flag would require some knowledge from the user about topology. I don't think such requirement should be imposed on someone who could use this piece of code.

I guess it would be ideal to pass input nodes to the `elementwise_add` pattern and let the pattern figure out binding between arguments.

Reply:
I think that if this was ok: https://github.com/PaddlePaddle/Paddle/blob/d0000082c4b88102bed62e8165b503f0e0481bac/paddle/fluid/framework/ir/graph_pattern_detector.h#L408
Then it would be also fine to use this flag for elementwise add pattern

Reply:
I would like to do such refactoring in a separate PR, if that's fine for you.

Reply:
Sure